sdks/python/apache_beam/transforms/util.py

```diff
@@ -864,3 +867,225 @@ def add_window_info(element, timestamp=DoFn.TimestampParam,
 
     def expand(self, pcoll):
       return pcoll | ParDo(self.add_window_info)
+
+
+class Regex(object):
+  """
+  PTransform  to use Regular Expression to process the elements in a
+  PCollection.
+  """
+  @staticmethod
+  def _regex_compile(regex):
+    """Return re.compile if the regex has a string value"""
+    if isinstance(regex, str):
+      regex = re.compile(regex)
+    return regex
+
+  @staticmethod
+  @ptransform_fn
+  def _matches_all_object(pcoll, regex, group=None):
+    """
+    PTransform to find the matches as per the regex.
+    """
+    def _process(element):
+      m = regex.finditer(element)
+      results = list()
+      for _, match in enumerate(m, start=1):
+        results.append(match.group())
+        for groupNum in range(0, len(match.groups())):
+          results.append(match.group(groupNum + 1))
+      if results:
+        yield results
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @ptransform_fn
+  def _matches_kv_object(pcoll, regex, keyGroup, valueGroup):
+    """
+    PTransfrom to find the matches as per regex and return as a KV pair.
+    """
+    def _process(element):
+      match = regex.match(element)
+      if match:
+        yield (match.group(keyGroup), match.group(valueGroup))
+
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def matches(pcoll, regex, group=None):
+    """
+    Returns the matches if the entire line matched the Regex. Returns the
+    entire line (group 0 by default). Group can be integer value and a string
+    value.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+      group: (optional) name of the group, it can be integer or a string value.
+    """
+    regex = Regex._regex_compile(regex)
+    group = group or 0
+
+    def _process(element):
+      m = regex.match(element)
+      if m:
+        yield m.group(group)
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
```

Comment:
Regex._matches_all_object returns a PCollection of lists (of groups). I'm not, however, sure that's what one wants. 

It seems the most primitive operation would be to have a transform that returns the (non-trivial) match objects themselves. Everything could then be implemented on top of that (e.g. extracting various groups). 

There's also a question of whether, when doing a "match all", a PCollection of lists (on for each matching element) would be best to return, or a flattened pcollection of all matches from all elements (less information, so I'd defer to the former or let it be optional).

Reply:
As per the java implementation, matches_kv return the list of KV (of groups). 
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Regex.java#L558 
I've updated the code little bit and found we no longer require `Regex._matches_all_object` as a separate private method so I moved the functionality in the `Regex.matches_kv` and removed the `Regex._matches_all_object` method.
(Previously I was using `Regex._matches_all_object` for both (`matches_kv` & `find_kv`), and now after the changes, I am using `re.match` for `matches_kv` and `re.finditer` for `find_kv`.)

And the third point regarding the return of the "match all". I think (the first approach you mentioned) PCollection of lists (containing the matching element of each item) is more relevant.  Because (as in the java api) "match all" applies to a single item of a collection, not to the entire collection.